### DISCUSSION TITLE
BT81x in micropython for ESP32

### DISCUSSION THREAD

**Discussion by Thompie2000** [RudolphRiedel/FT800-FT813#43] [Q&A]
hello

I am currently writing micropython for bt81x. i am using ESP32 as microcontroller. I got the code based from RudolphRiedel. I have already written a piece of code, but I cannot get past the REG_ID. Can you help me with this or do you have a micropython librarie?

Kind regards
Thompie2000

[Riverdi_library.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8572102/Riverdi_library.zip)

**Comment by RudolphRiedel** [comment]
I am entry level bad with Python.
But the first thought I had was there must be a way to use C-Libraries with this.
Which on a second thought would require to compile the whole thing instead of just throwing a bunch of .py files at it.

Do you have a Logic-Analyzer? First make sure the first steps are working correctly, like getting the bytes in the correct order and sending all of them.

    host_cmd(0x00)
    host_cmd(0x68)
    host_cmd(CLK_EXT)
    host_cmd(CLKSEL)
    host_cmd(ACTIVE)

The first one has to go, this also is ACTIVE.

    time.sleep(0.3)

This is seconds, this could be changed to time.sleep_ms(50).

    while(b'7C' != rd8(REG_ID)):

And now my Python left me, maybe.
But what does b'7C' really mean? The bit value of the string "7C"?

Why is this not: while(0x7c != rd8(REG_ID): ?

        print("waiting")
        time.sleep(0.1)

And this could be changed to time.sleep_ms(1).

And now I am curious what the SPI actually looks like up this point.

**Comment by Thompie2000** [comment]
best


Thank you very much for your prompt reply. I have modified the code and also send the result through the logic analyzer.

Kind regards
Thompie2000

[Riverdi3.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8580358/Riverdi3.zip)

**Reply by RudolphRrr** [reply]
Please attach a .sal capture file for the whole tranfer.
However, the "logic_analyzer REG_ID.png" shows an issue, not only should there be 3 bytes since the address is 3 bytes long, the first byte should be 0x30, so your order is not correct.
Note that the address uses a different byte-order than the data following the address.

So a complete read-transaction for a single byte is 5 bytes, for REG_ID:
0x30 0x20 0x00 0x00 0x00

And the expected answer as seen in the logic-analyzer is:
0x00 0x4a 0x43 0x42 0x7c

**Comment by Thompie2000** [comment]
best

Now I see the problem. I am now getting further with REG_ID and REG_CPURESET.

Can you please check if the WR32 , WR16 and WR8 are okay?

Kind regards
Thompie2000

[Riverdi3.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8582168/Riverdi3.zip)
s

**Comment by RudolphRrr** [comment]
This is starting to look good on the Logic Analyzer, my Python is still not good. :-)
Please comment out the print() lines in the transfer functions and set the sample rate much higher, 2 MS/s does not really work with a clock of 1MHz, make it at least 10MS/s or even higher.
The analyzer has some issues with the trace now, MOSI is not decoded and it looks like there are some chip-select changes missing.

**Reply by Thompie2000** [reply]
The reason why I use 1MHz is because when I use 2MHz my logic analyzer doesn't start working very well

**Reply by Thompie2000** [reply]
Is it possible that I may see your logic analyzer during initialisation? That way, I can also see where I am going wrong.

Thank you very much for your help.

kind regards
Thompie2000

**Reply by RudolphRrr** [reply]
Oh wait, are you using a clone? If this is the case using Sigrok/Pulseview might be the better option.
2MHz is not fast enough to sample the 1MHz signal, not when looking at details.

And yes, I can provide a logfile, this has to wait untill I am back home though. :-)

**Reply by RudolphRiedel** [reply]
I just took a trace from my current project.
This is from powerup thru init, writing of the images, checking touch to writing the first display list.  
[SPI_Trace_BT81x_Startup.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8584186/SPI_Trace_BT81x_Startup.zip)

Channel 0 is the PD line.
The first thing that happens is CS going high, this is right after reset of the controller.
PD is initialized to low and stays low for 6ms.
After PD is put to high there is a 21ms delay before the first commands are issued to start the EVE chip.
Then I have a 40ms delay before the chip ID is read and on the second attempt it reads back as 0x7c.
Well, and so on. :-)

**Comment by Thompie2000** [comment]
I am a small step ahead now. Now I would like to do the getting started (see photo). But unfortunately, I get nothing on the screen. Maybe I am doing something wrong when shifting. I am not very good at shifting.

Thank you for taking the time.

Kind regards
Thompie2000
[Riverdi3.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8589075/Riverdi3.zip)

**Reply by RudolphRrr** [reply]
The trace does not seem to match the code and probably due to the low sample rate quite a few of the chip-select transitions are missing.

Disable channel 4 to 7 and try again to increase the sample rate, please.

From what little I can gather from the Riverdi3.py, line 110 with "wr8(REG_DLSWAP,DLSWAP_FRAME)" should not be commented out,
the EVE chip needs the mostly empty display list to work with when the clock is enabled.

And you need that line at the end of the test() function or the new display list never gets drawn.

Which display do you have?
Micropython supports PiPico boards and I have one.

**Reply by Thompie2000** [reply]
I use the screen called RVT50AQBFWC00

**Reply by RudolphRrr** [reply]
Ok, so the EVE3-50G I am using is pretty close, at least in regards of displaying anything.
The EVE3-50G is like the RVT50UQBNWC00 but uses a different touch controller.

Well, I can't try anything here right now.

**Reply by Thompie2000** [reply]
and for the text to appear on the screen. Is that similar to your controller and my controller?

**Reply by RudolphRrr** [reply]
It's a BT815 with 800x480  in 5", only the timing parameters might be a little different.
As for the controller, that is something MicroPython has to deal with. :-)

**Comment by Thompie2000** [comment]
is it also possible that you can send the program code? That way I can also look at the recordings of your logic analyzer and also your program code side by side. Maybe I will then see my mistake?

Thanks again very much for your time

Kind regards
Thompie2000

**Reply by RudolphRiedel** [reply]
Well, the code is exactly what I have in this repository, at least in regards of the initialization.
Here is annother trace I took yesterday, this time with a Raspberry Pi Pico board running bare-metall with a library version from last year:
[RP2040_EVE3-50G.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8596505/RP2040_EVE3-50G.zip)

Now I have MicroPython on the Pico and I am working thru the confusion.
I am using PyMakr with VSCode and I have a main.py now that gets automatically started because of it's name and that blinks the LED and toggles the CS pin.

**Reply by RudolphRiedel** [reply]
...and it looks like I locked myself out of the board, can't connect anymore with main.py in place. :-(

**Comment by RudolphRiedel** [comment]
It's not like Python and I are friends anytime soon, but I dug in a little deeper and have test-code for you.
[Micropython_RP2040_EVE_SPI_test.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8597383/Micropython_RP2040_EVE_SPI_test.zip)

All this does is go thru basic initialization while assuming there is a BT81x as target with an external crystal.
As you can see in the attached trace, the commands at the beginning are the same as in my earlier trace, then the wait for 0x7c from REG_ID is successfull, reading of REG_CPURESET yields 0x02 at first and finally 0x00.
Then I use the wr8(), wr16() and wr32() functions to write test-values and in the loop I read these back.

Well, the basic functionality is there but as much fun as this exercise provided, I rather stick with C. :-)

**Reply by Thompie2000** [reply]
Thank you very much for the basic code. Hopefully I can now continue programming from here. I think I got stuck on the basic code. 

Thanks again for your time.

Kind regards
Thompie2000

**Comment by Thompie2000** [comment]
The initialisation seems to work now thanks to your code. But when I want to do the getting started (see attachment) as it says on the datasheet, I can't do it. Am I still doing the wrong thing in shifting or not?

You are really helping me here. Thank you very much.

Kind regards
Thompie2000

https://brtchip.com/wp-content/uploads/Support/Documentation/Programming_Guides/ICs/EVE/BRT_AN_033_BT81X_Series_Programming_Guide.pdf

[Riverdi4.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8601728/Riverdi4.zip)

**Reply by RudolphRiedel** [reply]
I got hit with a cold, drinking lots of tea kind of hold me back on getting it to work so far to actually display something. :-)

And I see this as an opportunity to do more Python, it helps with learning programming languages when you have a good idea on what you like to do. :-)
So that I implemented this mostly on my own instead of trying to correct your code is actually helping me. :-)
And it was a pain to find out how to actually use the SPI functions.
Fun.

**Reply by RudolphRiedel** [reply]
And just to contradict myself, I went with your functions after implementing the rest of eve_init(). :-)

[Riverdi4.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8602497/Riverdi4.zip)

It displays the inital screen here - refreshed about every 20ms.
Also switching to 8MHz SPI works here.

The pin configuration is in line 75ff.

The only issue was in your clear() function, the parameters for this one are bits, not bytes.

The timing parameters might be a little off for your display.

And put in your own license when you release this later! :-)

**Comment by Thompie2000** [comment]
best of luck

I have now been able to set up my first test display. But when I switch off my display and switch it on again a little later, my panel display no longer works. If I turn off my display a little later. At the moment it goes off I see that the panel display goes on and off for a moment. What can be wrong?

Kind regards
Thompie2000

**Reply by RudolphRrr** [reply]
What exactly do you mean by "turn off" and "switch it on again later"?
By software or by hardware?

**Reply by Thompie2000** [reply]
By turning hardware on and off

**Reply by RudolphRrr** [reply]
Do you actually use the PD line?
Or more precisely, how do you connect everything?

I am using this board:
https://github.com/RudolphRiedel/EVE_display-adapter/tree/master/L-D5019-08/L-D5019-08-01

So I am running a bunch of jumper wires to the RP2040:
![grafik](https://user-images.githubusercontent.com/75015332/166416630-33aa2af9-4dc8-4160-9f54-bc0c7257e0e0.png)

VCC is connected to an extra PSU which I have currently set to 7V,
And GND is of course connected to both the PSU and the RP2040 board.

Edit; this means I am using 3.3V for the backlight as I need to with the EVE3-50G, but the RVT50AQBFWC00 would work with 5V for the backlight

**Reply by Thompie2000** [reply]
I have connected most of the connections via ESP32. For the power supply, I connected it via the power source of +3V3.

![Break_board-20-maxi-700x561](https://user-images.githubusercontent.com/75269299/166421233-8b42581c-19a0-4ff5-b1b6-d530bb863802.png)

**Reply by RudolphRrr** [reply]
Where do the 3.3V come from? From the ESP32 board? That might be a bit much for the board.
This does not explain though why it does not always start.

**Reply by Thompie2000** [reply]
For +3V3 I have connected to the lab power supply

**Reply by RudolphRrr** [reply]
Hmm.
And just to be clear, the issue is not that the python script is not automatically started?

Remote diagnosis, always fun, always a fine line between the right questions and stupid questions. :-)

**Reply by Thompie2000** [reply]
They can still communicate with each other when I look through the logic analyzer, but the Panel display no longer works. Even when I set the Pin to low.

**Reply by RudolphRrr** [reply]
I usually only get this issue with the PD pin not connected and resetting the controller.
Without the PD pin connected the display is not properly initialized.

Try adding a  host_cmd(0x68, 0x00) in eve_init() between time.sleep(21) and host_cmd(CLK_EXT, 0x00).

**Reply by Thompie2000** [reply]
I think I have found the culprit. The lab power supply that I use is so old that it no longer functions properly. When I turn off the lab power supply it goes to a very high voltage and discharges for a very long time. So unfortunately I have to buy a new screen.

**Reply by RudolphRiedel** [reply]
Yikes, equipment failure is the worst.

**Comment by Thompie2000** [comment]
Hello

I have now bought a new screen and have been working on my test function. The indicated place when I press on something and displays the coordinates works well. I would now like to try my first test function. On the screen should be a red dot with text. I have done it this way (see appendix line 250). But unfortunately nothing appears on the screen. Am I doing something wrong?

Kind regards
Thompie2000
[Riverdi_library.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8660386/Riverdi_library.zip)

**Reply by RudolphRiedel** [reply]
I had a quick glimpse and your clear() function is broken, please check this version I posted last week:
https://github.com/RudolphRiedel/FT800-FT813/files/8602497/Riverdi4.zip

**Comment by Thompie2000** [comment]
Best

I have set a current limiter on my lab power supply at 0.440A. But I forgot about the current peaks that cause my display to go on and off very quickly and therefore my display stopped working at some point. Is it possible to repair this or not?

Kind regards
Thompie2000

**Reply by RudolphRrr** [reply]
Well, maybe.
But that is something I have no answer for.
I would offer to check the PSU but it is rather unlikely that you are from the same area and shipping is not really an option when you can get a good bench PSU for less than 100 EUR.
I am using a Peaktech 6225A here at home on my desk and I see it currently listed at around 70 EUR.

Ok, and I am usually not using the output of the PSU directly, all my boards have regulators on them.

**Comment by Thompie2000** [comment]
I'm almost done with what I would like to have. I am currently writing function from CMD_text, apparently something is still wrong with the private_string_write and I don't see direct the error. 

Thank you for taking the time to do this.

Kind regards
Thompie2000
[Riverdi5.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8706218/Riverdi5.zip)

**Reply by RudolphRiedel** [reply]
I decided to give it a spin to the check what the logic analyzer reports for the command.
But if I am not mistaken the issue is something else.

There are actually 3 command interfaces.
First is the host commands, this works fine.
Second is writing display list commands directly to the display list.
My library does not do that except for the inital list in the init function.

Third is where the action is, using the command co-processor to build the display list and this is what you can find in my library.
The "EVE_cmd_xx" functions.

Building a simple list looks like this:
    EVE_cmd_dl(CMD_DLSTART);
    EVE_cmd_dl(DL_CLEAR_RGB | WHITE);
    EVE_cmd_dl(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG);
    EVE_cmd_dl(DL_COLOR_RGB | BLACK);
    EVE_cmd_text(50, 50, 28, 0, "Hello there!");
    EVE_cmd_dl(DL_DISPLAY);
    EVE_cmd_dl(CMD_SWAP);
    EVE_execute_cmd();

EVE_cmd_dl() is for the most basic commands that require no parameters,  this is a single 32 bit word.
Some are co-processor commands, like CMD_DLSTART or CMD_SWAP, others are are display list commands that are
just copied into the new display list the co-processor builds for us. 

So that would be:

def cmd_dl(command)
    eve_begin_cmd(command)
    cs(1)

Which I just derived from what you already had.

Translating the above with what you already have in place I got to this:
def test():
    cmd_dl(0xFFFFFF00) # CND_DLSTART
    cmd_dl(0x02000000 + 0x00FFFFFF) # DL_CLEAR_RGB + WHITE
    cmd_dl(clear(1, 1, 1))
    cmd_dl(color_rgb(0, 0, 0))
    cmd_text(100, 200,31,0,"Hello there!")
    cmd_dl(0x00000000) # DL_DISPLAY
    cmd_dl(0xFFFFFF01) # CMD_SWAP

This is missing EVE_execute_cmd().

And this is my function:
void EVE_execute_cmd(void)
{
    while (EVE_busy()) {};
}

So this requires a busy() function and there is none so far.

So for the test if this works, if I somehow manage again to get this over to the RP2040,
I'll be just using this:

time.sleep(2)
print("trigger Logic Analyzer")
time.sleep(2)

eve_init()
test()
while(1):

**Reply by Thompie2000** [reply]
i put the code as you had described in my code and ran it once. I am now having a problem with my private_string_write in particular. The compiler says that the string index is out of range.

The code looks like this.

def cmd_text(x,y,font,opt,text):
    
    eve_begin_cmd(CMD_TEXT)
    buf = bytearray([0,0,0,0,0, 0])
    buf[0] = x
    buf[1] = x >> 8
    buf[2] = y
    buf[3] = y >> 8
    buf[4] = font
    buf[5] = font >> 8
    buf[6] = opt
    buf[7] = opt >> 8
    spi.write(buf)
    private_string_write(text)
    cs(1)
    
def private_string_write(text):
     textindex = 0
     byte = text
     padding = 0
     while((byte[textindex] != 0) and (textindex < 249)):
         spi.write(byte[textindex])
         textindex = textindex + 1
     padding = textindex & 3
     padding = 4- padding
     while(padding > 0):
         spi.write(0)
         padding = padding - 1 

Can you help me with this?

Kind regards
Thompie2000

**Reply by RudolphRrr** [reply]
I did not manage to test it yesterday, after several tries to get PyMakr to work and the console just not connecting to the target I had to give up.
Any advice on how to write the .py file to the MicroPython target? I find this highly frustrating.

Looking at the function now, spi,write uses a bytearray.
So spi,write(byte[textindex]) would not send a single char but everything from this position.

Well, what actually is important is that the bytes sent ad up to full 32 bits, so the padding at the end matters.

def private_string_write(text):
 encoded_string = text.encode() # encode to UTF-8
 byte = bytearray(encoded_string)
 spi.write(byte)
 count = len(byte)
 padding = count & 3
 padding = 4 - padding
 while(padding > 0):
  spi.write(0)
  padding = padding - 1
 
This might work but I only hacked this together on the fly with help from the internet and "tested" the string conversion and padding portion of this with an online python compiler.

**Reply by Thompie2000** [reply]
Unfortunately, I now get a black screen. Did I forget to set a command somewhere?

def test():
    cmd_dl(0xFFFFFF00) # CND_DLSTART
    cmd_dl(0x02000000 + 0x00FFFF) # DL_CLEAR_RGB + WHITE
    cmd_dl(clear(1, 1, 1))
    cmd_dl(color_rgb (0, 0, 0))
    cmd_text(100, 200,31,0, "Hello World!")
    cmd_dl(0x000000) # DL_DISPLAY
    cmd_dl(0xFFFFFF01) # CMD_SWAP

def cmd_text(x,y,font,opt,text):
    
    eve_begin_cmd(CMD_TEXT)
    buf = bytearray([0,0,0,0,0, 0])
    buf[0] = x
    buf[1] = x >> 8
    buf[2] = y
    buf[3] = y >> 8
    buf[4] = font
    buf[5] = font >> 8
    buf[6] = opt
    buf[7] = opt >> 8
    spi.write(buf)
    private_string_write(text)
    cs(1)
    
def private_string_write(text):
    encoded_string = text.encode() # encode to UTF-8
    byte = bytearray(encoded_string)
    spi.write(byte)
    count = len(byte)
    padding = count & 3
    padding = 4 - padding
    while(padding > 0):
        spi.write(bytearray([0]))
        padding = padding - 1

**Reply by RudolphRrr** [reply]
I can't really tell right now and at home I still would need to find a way to put this on the board.
I could take a guess when you provide a trace.

You could try to add  something else like a red dot.

**Reply by Thompie2000** [reply]
When i use this function. This function work perfect. But when i use bloeddruk(). It will not work. When i use bloeddruk function the right screen is full white and the left screen i see a few letters. The letter that i see on the screen is, 
VOLG DE I
OP DE BL

What is wrong with bloeddruk function?

def choose_lang():
    wr32(RAM_DL + 0, clear(1, 1, 1))
    wr32(RAM_DL + 4, begin(BITMAPS))                       # start drawing bitmaps
    wr32(RAM_DL + 8,  vertex2II(200, 200, 31, 'E'))         # ASCII T in font 31
    wr32(RAM_DL + 12, vertex2II(234, 200, 31, 'N'))         # ASCII E in font 31
    wr32(RAM_DL + 16, vertex2II(268, 200, 31, 'G'))        # ASCII X in font 31
    
    wr32(RAM_DL + 20, vertex2II(331, 200, 31, 'F'))         # ASCII T in font 31
    wr32(RAM_DL + 24, vertex2II(364, 200, 31, 'R'))         # ASCII E in font 31
    
    wr32(RAM_DL + 28, vertex2II(434, 200, 31, 'N'))         # ASCII T in font 31
    wr32(RAM_DL + 32, vertex2II(468, 200, 31, 'L'))         # ASCII E in font 31
    
    wr32(RAM_DL + 36, vertex2II(100, 300, 31, 'B'))         # ASCII T in font 31
    wr32(RAM_DL + 40, vertex2II(134, 300, 31, 'A'))         # ASCII E in font 31
    wr32(RAM_DL + 44, vertex2II(168, 300, 31, 'C'))         # ASCII T in font 31
    wr32(RAM_DL + 48, vertex2II(202, 300, 31, 'K'))         # ASCII E in font 31
    
    wr32(RAM_DL + 52, vertex2II(400, 300, 31, 'N'))         # ASCII T in font 31
    wr32(RAM_DL + 56, vertex2II(434, 300, 31, 'E'))         # ASCII E in font 31
    wr32(RAM_DL + 60, vertex2II(468, 300, 31, 'X'))         # ASCII T in font 31
    wr32(RAM_DL + 64, vertex2II(502, 300, 31, 'T'))         # ASCII E in font 31
    
    wr32(RAM_DL + 68, end())
    wr32(RAM_DL + 72, display())                           # display the image
    wr32(REG_DLSWAP, 2) # tell EVE to swap the display lists which makes this one active

def bloeddruk():
    wr32(RAM_DL + 0, clear(1, 1, 1))
    wr32(RAM_DL + 4, begin(BITMAPS))                       # start drawing bitmaps
    wr32(RAM_DL + 8,  vertex2II(100, 200, 31, 'V'))         # ASCII T in font 31
    wr32(RAM_DL + 12, vertex2II(134, 200, 31, 'O'))         # ASCII E in font 31
    wr32(RAM_DL + 16, vertex2II(168, 200, 31, 'L'))        # ASCII X in font 31
    wr32(RAM_DL + 20, vertex2II(202, 200, 31, 'G'))         # ASCII T in font 31
    wr32(RAM_DL + 24, vertex2II(236, 200, 31, ' '))         # ASCII E in font 31
    wr32(RAM_DL + 28, vertex2II(270, 200, 31, 'D'))         # ASCII T in font 31
    wr32(RAM_DL + 32, vertex2II(304, 200, 31, 'E'))         # ASCII E in font 31
    wr32(RAM_DL + 36, vertex2II(338, 200, 31, ' '))         # ASCII E in font 31
    wr32(RAM_DL + 40, vertex2II(362, 200, 31, 'I'))         # ASCII T in font 31
    wr32(RAM_DL + 44, vertex2II(434, 200, 31, 'N'))         # ASCII E in font 31
    wr32(RAM_DL + 48, vertex2II(468, 200, 31, 'S'))         # ASCII T in font 31
    wr32(RAM_DL + 52, vertex2II(502, 200, 31, 'T'))         # ASCII E in font 31
    wr32(RAM_DL + 56, vertex2II(536, 200, 31, 'R'))         # ASCII E in font 31
    wr32(RAM_DL + 60, vertex2II(570, 200, 31, 'U'))         # ASCII E in font 31
    wr32(RAM_DL + 64, vertex2II(604, 200, 31, 'C'))         # ASCII E in font 31
    wr32(RAM_DL + 68, vertex2II(638, 200, 31, 'T'))         # ASCII E in font 31
    wr32(RAM_DL + 72, vertex2II(672, 200, 31, 'I'))         # ASCII E in font 31
    wr32(RAM_DL + 76, vertex2II(706, 200, 31, 'E'))         # ASCII E in font 31
    wr32(RAM_DL + 80, vertex2II(0, 300, 31, ' '))         # ASCII E in font 31
    wr32(RAM_DL + 84, vertex2II(100, 300, 31, 'O'))         # ASCII E in font 31
    wr32(RAM_DL + 88, vertex2II(134, 300, 31, 'P'))         # ASCII E in font 31
    wr32(RAM_DL + 92, vertex2II(168, 300, 31, ' '))         # ASCII E in font 31
    wr32(RAM_DL + 96, vertex2II(202, 300, 31, 'D'))         # ASCII E in font 31
    wr32(RAM_DL + 100, vertex2II(236, 300, 31, 'E'))         # ASCII E in font 31
    wr32(RAM_DL + 104, vertex2II(270, 300, 31, ' '))         # ASCII E in font 31
    wr32(RAM_DL + 108, vertex2II(304, 300, 31, 'B'))         # ASCII E in font 31
    wr32(RAM_DL + 112, vertex2II(338, 300, 31, 'L'))         # ASCII E in font 31
    wr32(RAM_DL + 116, vertex2II(372, 300, 31, 'O'))         # ASCII E in font 31
    wr32(RAM_DL + 120, vertex2II(406, 300, 31, 'E'))         # ASCII E in font 31
    wr32(RAM_DL + 124, vertex2II(440, 300, 31, 'D'))         # ASCII E in font 31
    wr32(RAM_DL + 128, vertex2II(474, 300, 31, 'D'))         # ASCII E in font 31
    wr32(RAM_DL + 132, vertex2II(508, 300, 31, 'R'))         # ASCII E in font 31
    wr32(RAM_DL + 136, vertex2II(542, 300, 31, 'U'))         # ASCII E in font 31
    wr32(RAM_DL + 140, vertex2II(576, 300, 31, 'K'))         # ASCII E in font 31
    wr32(RAM_DL + 144, vertex2II(610, 300, 31, 'M'))         # ASCII E in font 31
    wr32(RAM_DL + 148, vertex2II(644, 300, 31, 'E'))         # ASCII E in font 31
    wr32(RAM_DL + 152, vertex2II(678, 300, 31, 'T'))         # ASCII E in font 31
    wr32(RAM_DL + 156, vertex2II(712, 300, 31, 'E'))         # ASCII E in font 31
    wr32(RAM_DL + 160, vertex2II(746, 300, 31, 'R'))         # ASCII E in font 31
    
    wr32(RAM_DL + 164, end())
    wr32(RAM_DL + 168, display())                           # display the image
    wr32(REG_DLSWAP, 2) # tell EVE to swap the display lists which makes this one active

**Reply by Thompie2000** [reply]
the problem I think is the window of the screen. When I shift my letter over 500 pixels to the right. Do I get to see a part white screen. Is that probably because of the initialization? I have the Hsize set to 800 and Vsize set to 480. Is there anything else I need to look at that?

**Reply by RudolphRrr** [reply]
I missed this detail,  VERTEX2II only works in a range of 0...511, VERTEX2F goes the full range but does not have the CELL parameter.
So with VERTEX2F you need an additional CELL command.

**Reply by RudolphRiedel** [reply]
I installed Thonny.
I do not like it but it gets the job done of copying the .py to the board.

So I played some more and now CMD_TEXT is working:
[Riverdi5.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8720775/Riverdi5.zip)

**Comment by RudolphRiedel** [comment]
I played with it some more:
[EVE_Test.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8748206/EVE_Test.zip)

 It has more of everything, more commands, display-list commands, some things are faster.

The touch-tag is read but somehow touch is not working so far, maybe because I did not write calibration values.

**Reply by Thompie2000** [reply]
wow what an improvement. Thank you for all your help. I am really learning a lot from you.

**Reply by RudolphRiedel** [reply]
Well, I hope that is limited to EVE because my Python is still bad. :-)
And getting (a little) better in Python is the only reason why I am even doing this, I am pretty sure this should be done as C-Library for Python and not directly in Python.

That touch is not working at all for me likely also is due to my lack of understanding of Python.
When I run the example for my EVE library as native code on the RP2040, it just works fine.

**Reply by RudolphRiedel** [reply]
Had to take a look again. :-)
[EVE_Test.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8757618/EVE_Test.zip)

And yes, touch actually was working, that I tried to use global variables inside functions without the "global" keyword did not work though.
So I learned a new keyword for Python today. :-)

And then I went a little further to discover that there are no static variables for Python functions.
But a class can have static variables, so I added a touch() class with a check() method.

The file is modified to run on your ESP32 setup.
The button behaves like a toggle switch now.
And the dot changes the color as long as it is touched.

**Reply by Thompie2000** [reply]
Thank you very much. It works very well now.  

I make my own PCB for my display and test if it works. the communication works well, the panel display turns on nicely, I can also use the touch, but... I get no display (No text and a background color) on the screen. The code is excactly the same and when I use without my own PCB it works very well. With my PCB it doesn't do so well. I managed to put it on a yellow background once, but after that, no more.

What could possibly go wrong?

The code now works 99% of the time. Only the display sleep mode and display wake up does it sometimes weird. Sometimes when it wakes up I get a white image and sometimes it is not a white image and I get to see the menu nicely.

Your code is really well done. I learned a lot about bit shifting there. Now I am starting to understand a little bit about the bit shifting of that

Kind regards
Thompie2000

**Comment by Thompie2000** [comment]
Thank you very much. It works very well now.

I make my own PCB for my display and test if it works. the communication works well, the panel display turns on nicely, I can also use the touch, but... I get no display (No text and a background color) on the screen. The code is excactly the same and when I use without my own PCB it works very well. With my PCB it doesn't do so well. I managed to put it on a yellow background once, but after that, no more.

What could possibly go wrong?

The code now works 99% of the time. Only the display sleep mode and display wake up does it sometimes weird. Sometimes when it wakes up I get a white image and sometimes it is not a white image and I get to see the menu nicely.

Your code is really well done. I learned a lot about bit shifting there. Now I am starting to understand a little bit about the bit shifting of that

Kind regards
Thompie2000

**Reply by RudolphRiedel** [reply]
Can you share the schematic for your board at least in parts?

>Your code is really well done.
No, it really isn't as that would imply that I know what I am doing in Python. :-)
I hacked my way thru this armed with the Internets and a Logic Analyzer.

And as much fun as this provided, doing this in Python is kind of crazy, the proper way to implement this would be as native library added to and compiled with MicroPython.

Which reminds me, I found something:
https://github.com/jamesbowman/CircuitPython_bteve

So there actually is an EVE library for Python, only for CircuitPython and not for MicroPython.

**Reply by Thompie2000** [reply]
here are my schmatics. Hopefully they are clear enough

[Schematic.zip](https://github.com/RudolphRiedel/FT800-FT813/files/8868473/Schematic.zip)

**Reply by RudolphRiedel** [reply]
I checked the schematic and found nothing wrong with it so far.
What type of inductors are L1 and L2 exactly?
And what type of capacitors are C6 and C11?

**Reply by Thompie2000** [reply]
The inductors are power inductors and the capacitor of C11 is Electrolytic Capacitor Radial leaded and capacitor of C6 is SMD aluminium Electrolytic Capacitor Radial can.

I have measured with the oscilloscope and the correct output power comes out.

**Reply by RudolphRiedel** [reply]
> The inductors are power inductors and the capacitor of C11 is Electrolytic Capacitor Radial leaded
> and capacitor of C6 is SMD aluminium Electrolytic Capacitor Radial can.

Yes, I assumed as much from the schematic but would like to know the exact make and model to confirm these are ok. :-)

> I have measured with the oscilloscope and the correct output power comes out.

Was this under load with the display initialized and backlight switched on?

Something must be different after all.

**Reply by Thompie2000** [reply]
The capacitor of C11 is:
Electrolytic Capacitor, 120 µF, 35 V, ± 20%, Radial Leaded, 2000 hours @ 105°C, Polar
**producer**: PANASONIC

The capacitor of C6 is:
SMD Aluminium Electrolytic Capacitor, Radial Can - SMD, 10 µF, 25 V, 3000 hours @ 105°C, Polar
**producer**: KEMET

The inductor of L1 is:
Power Inductor (SMD), 47 µH, 2.8 A, 4.7 A, ETQP, 10.7mm x 10mm x 4mm
**producer**: PANASONIC

the inductor if L2 is:
Power Inductor (SMD), Wirewound, 22 µH, 1.8 A, Shielded, 2 A, VLS-EX, 5mm x 5mm x 4.5mm
**producer**: TDK

**Reply by Thompie2000** [reply]
I put the excact working code on the PCB. I notice sometimes when I pull out the power supply and plug it back in and it goes on to initialization and the push buttons also work, but still no display.

**Reply by RudolphRiedel** [reply]
> Electrolytic Capacitor, 120 µF, 35 V, ± 20%, Radial Leaded, 2000 hours @ 105°C, Polar

EEUFC1V121?
117mR @100kHz/20°C might be a little high and 555mA ripple current might be a little low.

> The capacitor of C6 is:
> SMD Aluminium Electrolytic Capacitor, Radial Can - SMD, 10 µF, 25 V, 3000 hours @ 105°C, Polar
> **producer**: KEMET

This might be the problem and I missed it in the schematic earlier.
I meant to ask for C9 but C6 also is important.
Diodes lists vastly different values in ANP013_1509.pdf, 330µF for what you have as C6 and 1000µF for what you have as C9.
Wow, that is huge.
So switching on the backlight might cause a big voltage drop.

L1 is looking good.

L2 is a VLS5045EX-220M?
This is probably too small.
I am by no means an expert for switching regulators but the saturation current seems too low and the
DC resistance is rather high at 125mR.

**Reply by Thompie2000** [reply]
Indeed, I checked again with the scope and indeed see a voltage drop-out from de 3V3 and 5V. Probably it is indeed due to too low capacitors. 

the datasheet voor DC-DC converter is here:

https://www.ti.com/lit/ds/symlink/lm2651.pdf?HQS=dis-mous-null-mousermode-dsf-pf-null-wwe&ts=1655189146014&ref_url=https%253A%252F%252Fwww.mouser.be%252F
